04_Scheduling/04_Locks.md

```diff
@@ -0,0 +1,205 @@
+
+```c
+void acquire(spinlock_t *lock) {
+
+    while(1) {
+        if(lock->locked == false) {
+            lock->locked = true;
+            break;
+        }
+    }
+}
+```
+
+The release function is pretty simple, it just has to set the locked field to `false`: 
+
+```c
+void release(spinlock_t *lock) {
+    lock->locked = false;
+}
+```
+
+Consider the sequence of tasks outlined in the previous section, let's see what happens now using the `acquire/release`:
+
+* The lock is created and is set to false
+* The process A execute the `acquire` instruction so the locked variable is set to true.
+* Process A start to write the first letter to the shared resource
+* Process A is preempted and C started running, C is not using the shared resource so it goes doing whatever he was doing.
+* Process C is preempted and then B start running, B try to acquire the lock, but locked variable is already true, so it can't leave the loop in the acquire function.
+* Sooner or later the process B is preempted and C starts running again, it keep doing is job
+* C is preempted and it is time for B again now to run, that it is still in the acquire function loop, and since the lock is still set to true, it will be stuck there until the next preemption
+* Process B is preempted and now is time for A to run, it writes it second character 
+* ... skip until A hasn't finished to write it's string (B will be always kept waiting in the acquire loop)
+* A is running and has just sent the latest character to the shared resource, it leaves the string write loop, and run the `release` function that set the locked varialbe to false.
+* A is preempted and then sooner or later B will start running, now it is still in the acquire loop, but at the next iteration it will find the variable locked to false, so it can set it to true and finally leave the loop, and start write it's string...
+* ... etc.
+
+As we can see this simple algorithm has prevented two processes to interfere with each other. So can we assume that now we are safe? Well... no, there is one problem with the above implementation, this doesn't ensure mutual exclusion in a multi processor environment. Imagine we have 2 different processors calling the acquire at the same time (when we have a multi-cpu environment we have several processes/thread that run simultaneously) and they reach the if statement at the same time with the locked variable still to false, they will both reclaim the lock and will start to write to the shared resource, causing again a _race condition_. 
+
+But how to solve this issue? We need a real atomic operation that has to be used within the acquire loop. There are different solutions that we can adopt (for example use the `xchgl` instruction in the for loop, that is guaranteed to be atomic), but as usual we want to keep thing extremely simple, and in this case we can avail of the help of the compiler (gcc) that provides a whole set of atomic functions to be used by our kernel. 
```

Comment:
I wonder if it's worth adding a sentence here explaining what we mean by atomic?